### PR TITLE
model class to wrap around caffe networks

### PR DESCRIPTION
**Pull request by kashefy**
model class for feeding data into trained caffe neural nets.

### CODE REVIEW

File: src/+Models/CaffeModel.m [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        end
+        %% -----------------------------------------------------------------
+        
+    end
+    
+    %% -----------------------------------------------------------------------------------
+    methods(Access = protected)
+        
+        function [y,score] = applyModelMasked( obj, x )
+            blobs_in = x{1};
```
Comment (ivo--t) [inline src/+Models/CaffeModel.m:64]: I don't get where the cell is coming from? This method is called by the superclass, which hands over a matrix as x.

File: src/+Models/CaffeModel.m [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    % the LD_LIBRARY_PATH environment variable prior to launching MATLAB)
+    %% --------------------------------------------------------------------
+    properties (SetAccess = ?ModelTrainers.CaffeModel)
+        fpath_net_def;
+        fpath_weights;
+        thr;            % output node thresholds, default is 0.5
+        has_thr;
+    end
+    
+    properties (SetAccess = ?ModelTrainers.CaffeModel, Transient=false)
```
Comment (ivo--t) [inline src/+Models/CaffeModel.m]: Isn't false the default for Transient? Why the explicit mentioning?

File: src/+Models/CaffeModel.m [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            for ii = 1:numel(obj.net.outputs)
+                score.(obj.net.outputs{ii}) = double(blobs_out{ii});
+                d = blobs_out{ii};
+                if obj.has_thr
+                    thr_tmp = obj.thr{ii};
+                else
+                    thr_tmp = 0.5;
+                end
+                d(d >= thr_tmp) = 1;
+                d(d < thr_tmp) = -1;
```
Comment (ivo--t) [inline src/+Models/CaffeModel.m]: Just making sure: thr_tmp is always between 0 and 1, right?

File: src/+Models/CaffeModel.m (hunk 2 above)
Comment (kashefy) [reply src/+Models/CaffeModel.m]: this needs to be set to true, will do.

File: src/+Models/CaffeModel.m (hunk 1 above)
Comment (kashefy) [reply src/+Models/CaffeModel.m:64]: true, if featureMask in super class is not specified, x is handed over as is. In this case the KS calling applyModel( x ) defines x as a cell array, where each element is a feature matrix.
The network needs the individual features (ratemaps, ams...) as separate matrices, and the KS performs this separation.
this wouldn't work if featureMask defined in Model.Base is not empty.

File: src/+Models/CaffeModel.m (hunk 3 above)
Comment (kashefy) [reply src/+Models/CaffeModel.m]: yes, but a threshold outside the [0,1] range would also work.

Comment (ivo--t) [reply src/+Models/CaffeModel.m]: If thr_tmp could be above 1, line 107 could produce faulty decisions.

Comment (kashefy) [reply src/+Models/CaffeModel.m]: understood, will store a separate mask

**Comment by kashefy** [comment]
Current Status of this: revise new model class to behave more like other models (e.g. applyModel method  receives a feature vector and unrolls + reshapes it into the different feature groups required by caffe, this is currently done in the KS and this doesn't align well with the other model classes)

**Comment by kashefy** [comment]
@ivo--t , I've made some changes here. The CaffeModel class no longer inherits from Models.Base because, as you pointed out, it doesn't follow the same interface.
It would be nice to integrate this as a proper Model class but for now, I just need this for inference to be used by a KS. The KS for Id + loc as well as a variant of the DNNLocationKS that uses caffe instead of the usual matlab-only implementation (will make a pull request for this KS in wp3 repo very soon)

**Comment by kashefy** [comment]
Please review and let me know if anything is needed before we can merge this.

**Comment by ivo--t** [comment]
@kashefy , nothing is needed in order to merge this into the pipeline codebase. However, as 
-  at the moment it is not compliant with other +Models
-  as there is no ModelTrainer to produce this modeltype in the pipeline, 
-  as the pipeline could not use it for testing since the interface is different, 

I vote for not merging it into the pipeline but instead into the Blackboard-System code repo along with your respective KS that uses it. After all, that's (at the moment) the only purpose of this code.

If you agree: close this PR and create a new one in the BS repo, please.

**Comment by kashefy** [comment]
ok, will move this to the blackboard-system repo.

**Pull request closed without merging**